sandbox/modules/identity/src/main/java/org/opensearch/identity/authz/PermissionStorage.java

```diff
@@ -26,33 +29,57 @@
  */
 public class PermissionStorage {
 
-    public HashMap<Principal, List<Permission>> permissionStore = new HashMap<>();
+    public static HashMap<Principal, List<OpenSearchPermission>> permissionStore = new HashMap<>();
 
     /**
      * This function adds a set of permissions to the permission store. The principal is a unique identifier for some subject.
      * The List is a list of all permissions that are being granted.
      */
-    public void put(Principal principal, List<Permission> permissions) {
+    public static void put(Principal principal, List<OpenSearchPermission> permissions) {
 
-        permissionStore.put(principal, permissions);
+        for (OpenSearchPermission permission : permissions) {
+            permissionStore.computeIfAbsent(principal, permissionsList -> new ArrayList<>()).add(permission);
+        }
+    }
+
+    public static void put(String principal, List<OpenSearchPermission> permissions) {
+
+        StringPrincipal principalFromString = new StringPrincipal(principal);
+        for (OpenSearchPermission permission : permissions) {
+            permissionStore.computeIfAbsent(principalFromString, permissionsList -> new ArrayList<>()).add(permission);
+        }
     }
 
     /**
      * This function returns the List of permissions associated with the provided principal.
      * If permissions are modified during storage they must be reverted back to their original state during get().
      */
-    public List<Permission> get(Principal principal) {
+    public static List<OpenSearchPermission> get(Principal principal) {
 
         return permissionStore.get(principal);
     }
 
+    public static List<OpenSearchPermission> get(String principal) {
+
+        StringPrincipal principalFromString = new StringPrincipal(principal);
+        return permissionStore.get(principalFromString);
+    }
+
     /**
      * This function in-place deletes all targeted permissions associated with a given principal.
      */
-    public void delete(Principal principal, List<Permission> permissions) {
+    public static void delete(Principal principal, List<OpenSearchPermission> permissions) {
 
-        for (Permission permission : permissions) {
+        for (OpenSearchPermission permission : permissions) {
             permissionStore.remove(principal, permission);
         }
     }
+
+    public static void delete(String principal, List<OpenSearchPermission> permissions) {
+
+        StringPrincipal principalFromString = new StringPrincipal(principal);
+        for (OpenSearchPermission permission : permissions) {
+            permissionStore.remove(principalFromString, permission);
```

Comment:
Hi Craig thanks for the note. I am not sure I understand what you are saying here though. There are not any identifiers that would differ between two instances of the same permission or string principal. We don't have any differentiation between instances of two OpenSearchPermissions made from the same permission string. Too my knowledge there is no reason we cannot use this as is just like if we had a String : Int map. You don't have to override equals to evaluate whether two text patterns match because there is nothing that separates one instance of a string with matching characters from another. The same goes for the principals and permissions. 

You can see this working in #6220 in the integration tests. Unless I am misunderstanding what you are saying, this shows it works without override.

Reply:
Which test? `RestPermissionsIT.testPermissionsRestAPI` has this block:

```
// Delete the added permission
Request deleteRequest = new Request("DELETE", endpoint + "/" + username);
deleteRequest.setJsonEntity("{ \"permissionString\" : \"cluster.admin/read\"}\n");
Response deleteResponse = client().performRequest(deleteRequest);
assertThat(deleteResponse.getStatusLine().getStatusCode(), is(200));
```

but its not confirming that the permission was deleted, only that the API responded with 200.

Every class in Java extends from the parent Object class where the default `equals` method just sees if one instance of an object is that same instance as another.

If you define 2 OpenSearch permissions:

```
OpenSearchPermission osPerm1 = new OpenSearchPermission("cluster:monitor/health");
OpenSearchPermission osPerm2 = new OpenSearchPermission("cluster:monitor/health");

osPerm1.equals(osPerm2);  // This will return false unless `.equals` is overridden on OpenSearchPermission
```

Reply:
RestPermissionsIT and PermissionAPIIT both query the Permission Storage. They first assert that the permission is present and then assert that the permission is later no longer present.

Reply:
@scrawfor99 There's an error in the assertion. You are checking for a `String` in a `List<OpenSearchPermission>` so it will always return false. The assertion at the end of the test should be:

```
assertFalse(
    PermissionStorage.get((new StringPrincipal(username)))
        .stream()
        .map(permission -> permission.getPermissionString())
        .collect(Collectors.toList())
        .contains("cluster.admin/read")
);
```

to match the assertion above after you check to ensure the permission was added. After correcting the assertion, the test fails.

Reply:
Oof you are totally right. I will fix this.